src/odemis/acq/test/acq_test.py

```diff
@@ -564,6 +565,129 @@ def on_progress_update(self, future, start, end):
+        cls.cld = model.getComponent(role="cl-detector")
+        cls.ebd = model.getComponent(role="ebic-detector")
+
+    def test_valid_folding(self):
+        """
+        Test folding of 3 streams that have the same roi, dwellTime and repetition values but
+        different detector names
+        """
+        # Create the streams
+        sems = stream.SEMStream("test sem", self.sed, self.sed.data, self.ebeam,
+                                emtvas={"dwellTime", "scale", "magnification", "pixelSize"})
+
+        ebics = stream.EBICSettingsStream("test ebic", self.ebd, self.ebd.data, self.ebeam)
+
+        cls = stream.CLSettingsStream("test cl", self.cld, self.cld.data, self.ebeam)
+
+        sms_ebic = stream.SEMMDStream("sem ebic", [sems, ebics])
+        sms_cl = stream.SEMMDStream("sem cl", [sems, cls])
+
+        # SEM survey settings are via the current hardware settings
+        self.ebeam.dwellTime.value = self.ebeam.dwellTime.range[0]
+
+        cls.roi.value = (0, 0, 1, 1)
+        cls.repetition.value = (1024, 1024)
+
+        ebics.roi.value = (0, 0, 1, 1)
+        ebics.repetition.value = (1024, 1024)
+
+        folds = acqmng.foldStreams([sms_ebic, sms_cl])
+        folds = tuple(folds)
+        str_cnt = len(folds)
+
+        self.assertEqual(str_cnt, 1)
+        self.assertEqual(len(folds[0].streams), 3)
+
+
+    def test_folding_same_detector_name(self):
+        """
+        Test folding of the streams that have the same repetition and dwell time values but different
+        roi and detector names
+        """
+        # Create the streams
+        sems = stream.SEMStream("test sem", self.sed, self.sed.data, self.ebeam,
+                                emtvas={"dwellTime", "scale", "magnification", "pixelSize"})
+
+        ebics = stream.EBICSettingsStream("test ebic", self.cld, self.cld.data, self.ebeam)
+
+        cls = stream.CLSettingsStream("test cl", self.cld, self.cld.data, self.ebeam)
+
+        sms_ebic = stream.SEMMDStream("sem ebic", [sems, ebics])
+        sms_cl = stream.SEMMDStream("sem cl", [sems, cls])
+
+        # SEM survey settings are via the current hardware settings
+        self.ebeam.dwellTime.value = self.ebeam.dwellTime.range[0]
+
+        cls.roi.value = (0, 0, 1, 1)
+        cls.repetition.value = (1024, 1024)
+
+        ebics.roi.value = (0, 0, 1, 1)
+        ebics.repetition.value = (1024, 1024)
+
+        folds = acqmng.foldStreams([sms_ebic, sms_cl])
+        folds = tuple(folds)
+        str_cnt = len(folds)
+
+        self.assertEqual(str_cnt, 2)
+        self.assertEqual(len(folds[0].streams), 2)
+        self.assertEqual(len(folds[1].streams), 2)
+
+    def test_folding_pairs_of_two(self):
+        """
+        Test folding of the streams that have the same repetition and dwell time values but different
+        roi and detector names
+        """
+        # Create the streams
+        sems = stream.SEMStream("test sem", self.sed, self.sed.data, self.ebeam,
```

Comment:
Would it be nice to add the stream initialisation in the init as it is same for all the test cases or do we need flexibility to change them for future test cases?

Reply:
I don't think the sem streams would change in the future so moving it to the init is indeed valid. But my intention was to keep all the streams next to each other and felt the code is more readable that way.